File: lab1/lab1_1/Python/lab1_1.py
import tkinter as tk
from tkinter import messagebox

BG = "#f5f7fb"
CARD_BG = "#ffffff"
TEXT = "#202124"
MUTED = "#5f6368"
BORDER = "#dadce0"
PRIMARY = "#1a73e8"
PRIMARY_HOVER = "#1666d3"
DANGER = "#d93025"
OK = "#188038"

FONT = ("Segoe UI", 11)
FONT_B = ("Segoe UI", 11, "bold")
TITLE_FONT = ("Segoe UI", 16, "bold")


def _is_int_token(tok: str) -> bool:
    if tok.startswith("-"):
        return tok[1:].isdigit() and len(tok) > 1
    return tok.isdigit()


def format_int_with_commas(tok: str) -> str:
    raw = tok.replace(",", "")
    if not _is_int_token(raw):
        return tok  
    n = int(raw)
    return f"{n:,}"


def parse_two_ints(s: str):
    """
    Parse exactly 2 integers from a string like: "12 6" or "1,000 -2,500"
    Returns (a, b) as ints or raises ValueError with a friendly message.
    """
    parts = s.strip().split()
    if len(parts) != 2:
        raise ValueError("Please enter exactly TWO numbers separated by space.")
    p0 = parts[0].replace(",", "")
    p1 = parts[1].replace(",", "")
    if not _is_int_token(p0) or not _is_int_token(p1):
        raise ValueError("Numbers must be integers (e.g., 12 6).")
    return int(p0), int(p1)


class App(tk.Tk):
    def __init__(self):
        super().__init__()
        self.title("Product & Average Calculator")
        self.configure(bg=BG)
        self.resizable(False, False)

        self.card = tk.Frame(self, bg=CARD_BG, highlightthickness=1, highlightbackground=BORDER)
        self.card.pack(padx=20, pady=20)

        tk.Label(self.card, text="Compute Product & Average", bg=CARD_BG, fg=TEXT, font=TITLE_FONT)\
            .grid(row=0, column=0, columnspan=2, sticky="w", padx=18, pady=(16, 2))

        tk.Label(self.card, text="Enter two numbers per line (space-separated).",
                 bg=CARD_BG, fg=MUTED, font=FONT)\
            .grid(row=1, column=0, columnspan=2, sticky="w", padx=18, pady=(0, 14))

        # Input 1
        self._field_label("1st two numbers", row=2)
        self.in1, self.in1_placeholder = self._entry(row=3, placeholder="e.g., 12 6")
        self.msg1 = tk.Label(self.card, text="", bg=CARD_BG, fg=DANGER, font=("Segoe UI", 10))
        self.msg1.grid(row=4, column=0, columnspan=2, sticky="w", padx=18, pady=(0, 8))

        # Input 2
        self._field_label("Last two numbers", row=5)
        self.in2, self.in2_placeholder = self._entry(row=6, placeholder="e.g., 7 9")
        self.msg2 = tk.Label(self.card, text="", bg=CARD_BG, fg=DANGER, font=("Segoe UI", 10))
        self.msg2.grid(row=7, column=0, columnspan=2, sticky="w", padx=18, pady=(0, 8))


        # Buttons
        btn_row = tk.Frame(self.card, bg=CARD_BG)
        btn_row.grid(row=9, column=0, columnspan=2, sticky="ew", padx=18, pady=(6, 10))
        btn_row.grid_columnconfigure(0, weight=1)
        btn_row.grid_columnconfigure(1, weight=1)

        self.compute_btn = self._button(btn_row, "Compute", self.compute, primary=True)
        self.compute_btn.grid(row=0, column=0, sticky="ew", padx=(0, 8))

        self.clear_btn = self._button(btn_row, "Clear", self.clear, primary=False)
        self.clear_btn.grid(row=0, column=1, sticky="ew", padx=(8, 0))

        # Results
        self.results = tk.Frame(self.card, bg=CARD_BG, highlightthickness=1, highlightbackground=BORDER)
        self.results.grid(row=10, column=0, columnspan=2, sticky="ew", padx=18, pady=(6, 16))

        self.out1 = self._result_line("The product of 1st 2 numbers is:", 0)
        self.out2 = self._result_line("The product of 2nd 2 numbers is:", 1)
        self.out3 = self._result_line("The average of four numbers is:", 2)

        self.status = tk.Label(self.card, text="", bg=CARD_BG, fg=MUTED, font=("Segoe UI", 10))
        self.status.grid(row=11, column=0, columnspan=2, sticky="w", padx=18, pady=(0, 14))

        # Enter triggers compute
        self.bind("<Return>", lambda e: self.compute())

        # Live validation + auto-format
        self._wire_live_validation(self.in1, self.msg1, self.in1_placeholder)
        self._wire_live_validation(self.in2, self.msg2, self.in2_placeholder)

        self.in1.focus_set()

    # ---------------- UI helpers ----------------
    def _field_label(self, text, row):
        tk.Label(self.card, text=text, bg=CARD_BG, fg=TEXT, font=FONT_B)\
            .grid(row=row, column=0, columnspan=2, sticky="w", padx=18, pady=(0, 6))

    def _entry(self, row, placeholder=""):
        container = tk.Frame(self.card, bg=CARD_BG)
        container.grid(row=row, column=0, columnspan=2, sticky="ew", padx=18, pady=(0, 6))

        e = tk.Entry(
            container,
            bd=0,
            bg="#ffffff",
            fg=TEXT,
            font=FONT,
            insertbackground=TEXT,
            relief="flat",
            highlightthickness=1,
            highlightbackground=BORDER,
            highlightcolor=PRIMARY
        )
        # INNER padding so text isn't sagad sa border
        e.pack(fill="x", ipady=12, padx=8)

        ph = placeholder
        if ph:
            e.insert(0, ph)
            e.config(fg=MUTED)

            def on_focus_in(_):
                if e.get() == ph and e.cget("fg") == MUTED:
                    e.delete(0, "end")
                    e.config(fg=TEXT)

            def on_focus_out(_):
                if not e.get().strip():
                    e.insert(0, ph)
                    e.config(fg=MUTED)

            e.bind("<FocusIn>", on_focus_in)
            e.bind("<FocusOut>", on_focus_out)

        return e, ph

    def _button(self, parent, text, command, primary=False):
        bg = PRIMARY if primary else "#eef3fd"
        fg = "#ffffff" if primary else PRIMARY
        hover = PRIMARY_HOVER if primary else "#e3ecff"

        wrap = tk.Frame(parent, bg=CARD_BG, highlightthickness=1,
                        highlightbackground=(PRIMARY if primary else BORDER))
        label = tk.Label(wrap, text=text, bg=bg, fg=fg, font=FONT_B, padx=14, pady=10, cursor="hand2")
        label.pack(fill="both", expand=True)

        label.bind("<Enter>", lambda e: label.config(bg=hover))
        label.bind("<Leave>", lambda e: label.config(bg=bg))
        label.bind("<Button-1>", lambda e: command())
        return wrap

    def _result_line(self, text, row):
        line = tk.Frame(self.results, bg=CARD_BG)
        line.grid(row=row, column=0, sticky="ew", padx=12, pady=(10 if row == 0 else 6, 6))
        line.grid_columnconfigure(1, weight=1)

        tk.Label(line, text=text, bg=CARD_BG, fg=MUTED, font=FONT)\
            .grid(row=0, column=0, sticky="w")

        val = tk.Label(line, text="—", bg=CARD_BG, fg=TEXT, font=FONT_B)
        val.grid(row=0, column=1, sticky="e")
        return val

    # ---------------- Live validation + formatting ----------------
    def _get_clean_value(self, entry: tk.Entry):
        if entry.cget("fg") == MUTED:
            return ""
        return entry.get().strip()

    def _set_entry_border(self, entry: tk.Entry, ok: bool):
        entry.config(highlightbackground=(BORDER if ok else DANGER))

    def _validate_text(self, text: str):
        """
        Returns (ok: bool, reason: str).
        Live validation rules:
        - blank => not ok (but we'll show a gentle message only if user typed something)
        - must be exactly 2 tokens split by spaces
        - each token must be an integer (commas allowed)
        """
        if not text.strip():
            return False, "Enter two integers separated by a space."
        parts = text.strip().split()
        if len(parts) < 2:
            return False, "Need 2 numbers (example: 12 6)."
        if len(parts) > 2:
            return False, "Only 2 numbers are allowed on this line."
        p0 = parts[0].replace(",", "")
        p1 = parts[1].replace(",", "")
        if not _is_int_token(p0) or not _is_int_token(p1):
            return False, "Integers only (no letters/decimals)."
        return True, ""

    def _auto_format_entry(self, entry: tk.Entry):
        """
        Auto-format numbers with commas while typing.
        Example: "1000 2500" -> "1,000 2,500"
        Keeps it simple: after format, cursor moves to end (stable + reliable).
        """
        if entry.cget("fg") == MUTED:
            return
        raw = entry.get()
        if not raw.strip():
            return

        parts = raw.split()  # collapses multiple spaces; acceptable for this form
        if not parts:
            return

        # Only format up to first 2 tokens; keep extra tokens as-is (validation will catch it)
        formatted_parts = []
        for i, tok in enumerate(parts):
            if i < 2:
                formatted_parts.append(format_int_with_commas(tok))
            else:
                formatted_parts.append(tok)
        new_text = " ".join(formatted_parts)

        if new_text != raw.strip():
            entry.delete(0, "end")
            entry.insert(0, new_text)
            entry.icursor("end")

    def _wire_live_validation(self, entry: tk.Entry, msg_label: tk.Label, placeholder: str):
        def on_key_release(_):
            # Auto-format first so validation sees formatted text too
            self._auto_format_entry(entry)

            t = self._get_clean_value(entry)
            ok, reason = self._validate_text(t)

            # If placeholder/empty: keep neutral border + no loud message
            if not t:
                self._set_entry_border(entry, True)
                msg_label.config(text="", fg=DANGER)
                return

            self._set_entry_border(entry, ok)
            msg_label.config(text=reason, fg=DANGER)

        def on_focus_out(_):
            # On blur, format again (helps if paste happened)
            self._auto_format_entry(entry)
            t = self._get_clean_value(entry)
            if not t:
                self._set_entry_border(entry, True)
                msg_label.config(text="")
                return
            ok, reason = self._validate_text(t)
            self._set_entry_border(entry, ok)
            msg_label.config(text=reason, fg=DANGER)

        entry.bind("<KeyRelease>", on_key_release)
        entry.bind("<FocusOut>", on_focus_out)

    # ---------------- Actions ----------------
    def compute(self):
        s1 = self._get_clean_value(self.in1)
        s2 = self._get_clean_value(self.in2)

        self.status.config(text="", fg=MUTED)

        # Validate both first (and show borders/messages)
        ok1, r1 = self._validate_text(s1)
        ok2, r2 = self._validate_text(s2)

        if not s1:
            ok1, r1 = False, "First input is required."
        if not s2:
            ok2, r2 = False, "Second input is required."

        self._set_entry_border(self.in1, ok1 if s1 else False)
        self._set_entry_border(self.in2, ok2 if s2 else False)
        self.msg1.config(text=(r1 if not ok1 else ""))
        self.msg2.config(text=(r2 if not ok2 else ""))

        if not ok1 or not ok2:
            messagebox.showerror("Invalid Input", (r1 if not ok1 else r2))
            self.status.config(text="Fix the highlighted input(s).", fg=DANGER)
            return

        try:
            a, b = parse_two_ints(s1)
            c, d = parse_two_ints(s2)

            p1 = a * b
            p2 = c * d
            avg = (a + b + c + d) / 4
            avg_display = int(avg) if avg.is_integer() else avg

            self.out1.config(text=f"{p1:,}", fg=OK)
            self.out2.config(text=f"{p2:,}", fg=OK)
            # average might not be int; format nicely
            if isinstance(avg_display, int):
                self.out3.config(text=f"{avg_display:,}", fg=OK)
            else:
                self.out3.config(text=str(avg_display), fg=OK)

            self.status.config(text="Computed successfully ✔", fg=OK)

        except ValueError as e:
            messagebox.showerror("Invalid Input", str(e))
            self.status.config(text=str(e), fg=DANGER)

    def clear(self):
        for e in (self.in1, self.in2):
            e.delete(0, "end")
            e.event_generate("<FocusOut>")  # restores placeholder
            e.config(highlightbackground=BORDER)

        self.msg1.config(text="")
        self.msg2.config(text="")

        for o in (self.out1, self.out2, self.out3):
            o.config(text="—", fg=TEXT)

        self.status.config(text="", fg=MUTED)
        self.in1.focus_set()


if __name__ == "__main__":
    App().mainloop()

File: lab1/lab1_1/Java/lab1_1.java
import java.awt.*;
import java.awt.event.*;
import java.text.NumberFormat;
import java.util.Locale;
import javax.swing.*;
import javax.swing.border.*;
import javax.swing.event.DocumentEvent;
import javax.swing.event.DocumentListener;
import javax.swing.plaf.FontUIResource;
import javax.swing.text.*;

public class lab1_1 extends JFrame {

    // Colors (Google-ish)
    private static final Color BG = new Color(0xF5F7FB);
    private static final Color CARD_BG = Color.WHITE;
    private static final Color TEXT = new Color(0x202124);
    private static final Color MUTED = new Color(0x5F6368);
    private static final Color BORDER = new Color(0xDADCE0);
    private static final Color PRIMARY = new Color(0x1A73E8);
    private static final Color DANGER = new Color(0xD93025);
    private static final Color OK = new Color(0x188038);

    private final JTextField in1 = new JTextField();
    private final JTextField in2 = new JTextField();

    private final JLabel msg1 = new JLabel(" ");
    private final JLabel msg2 = new JLabel(" ");

    private final JLabel out1 = new JLabel("—");
    private final JLabel out2 = new JLabel("—");
    private final JLabel out3 = new JLabel("—");
    private final JLabel status = new JLabel(" ");

    private final Border normalBorder = new CompoundBorder(
            new LineBorder(BORDER, 1, true),
            new EmptyBorder(10, 12, 10, 12)
    );

    private final Border errorBorder = new CompoundBorder(
            new LineBorder(DANGER, 1, true),
            new EmptyBorder(10, 12, 10, 12)
    );

    private boolean formattingInProgress = false;

    public lab1_1() {
        super("Product & Average Calculator");
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setResizable(false);

        // Use a modern-looking font if available
        setUIFont(new Font("Segoe UI", Font.PLAIN, 13));

        JPanel root = new JPanel(new GridBagLayout());
        root.setBackground(BG);

        JPanel card = new JPanel(new GridBagLayout());
        card.setBackground(CARD_BG);
        card.setBorder(new CompoundBorder(new LineBorder(BORDER, 1, true), new EmptyBorder(16, 16, 16, 16)));

        GridBagConstraints c = new GridBagConstraints();
        c.gridx = 0; c.gridy = 0; c.weightx = 1; c.fill = GridBagConstraints.HORIZONTAL;

        JLabel title = new JLabel("Compute Product & Average");
        title.setForeground(TEXT);
        title.setFont(new Font("Segoe UI", Font.BOLD, 20));
        card.add(title, c);

        c.gridy++;
        JLabel subtitle = new JLabel("Enter two numbers per line (space-separated).");
        subtitle.setForeground(MUTED);
        subtitle.setFont(new Font("Segoe UI", Font.PLAIN, 13));
        c.insets = new Insets(6, 0, 14, 0);
        card.add(subtitle, c);
        c.insets = new Insets(0, 0, 0, 0);

        // Input 1
        c.gridy++;
        card.add(fieldLabel("1st two numbers"), c);

        c.gridy++;
        setupTextField(in1, "e.g., 12 6");
        card.add(in1, c);

        c.gridy++;
        setupMsgLabel(msg1);
        card.add(msg1, c);

        // Input 2
        c.gridy++;
        c.insets = new Insets(8, 0, 0, 0);
        card.add(fieldLabel("Last two numbers"), c);
        c.insets = new Insets(0, 0, 0, 0);

        c.gridy++;
        setupTextField(in2, "e.g., 7 9");
        card.add(in2, c);

        c.gridy++;
        setupMsgLabel(msg2);
        card.add(msg2, c);

        
        // Buttons
        c.gridy++;
        c.insets = new Insets(6, 0, 10, 0);
        JPanel btnRow = new JPanel(new GridLayout(1, 2, 12, 0));
        btnRow.setBackground(CARD_BG);

        JButton computeBtn = makeButton("Compute", true);
        JButton clearBtn = makeButton("Clear", false);

        btnRow.add(computeBtn);
        btnRow.add(clearBtn);
        card.add(btnRow, c);
        c.insets = new Insets(0, 0, 0, 0);

        // Results box
        c.gridy++;
        JPanel results = new JPanel(new GridBagLayout());
        results.setBackground(CARD_BG);
        results.setBorder(new CompoundBorder(new LineBorder(BORDER, 1, true), new EmptyBorder(10, 12, 10, 12)));

        GridBagConstraints r = new GridBagConstraints();
        r.gridx = 0; r.gridy = 0; r.weightx = 1; r.fill = GridBagConstraints.HORIZONTAL;

        results.add(resultRow("The product of 1st 2 numbers is:", out1), r);
        r.gridy++;
        r.insets = new Insets(6, 0, 0, 0);
        results.add(resultRow("The product of 2nd 2 numbers is:", out2), r);
        r.gridy++;
        results.add(resultRow("The average of four numbers is:", out3), r);

        card.add(results, c);

        // Status
        c.gridy++;
        c.insets = new Insets(10, 0, 0, 0);
        status.setForeground(MUTED);
        status.setFont(new Font("Segoe UI", Font.PLAIN, 12));
        card.add(status, c);

        // Root pack
        GridBagConstraints rc = new GridBagConstraints();
        rc.insets = new Insets(20, 20, 20, 20);
        root.add(card, rc);

        setContentPane(root);

        // Wire actions
        computeBtn.addActionListener(e -> compute());
        clearBtn.addActionListener(e -> clear());

        // Enter triggers compute from either field
        Action computeAction = new AbstractAction() {
            @Override public void actionPerformed(ActionEvent e) { compute(); }
        };
        in1.addActionListener(computeAction);
        in2.addActionListener(computeAction);

        // Live validation + auto-format
        wireLiveValidation(in1, msg1);
        wireLiveValidation(in2, msg2);

        pack();
        setLocationRelativeTo(null);
        in1.requestFocusInWindow();
    }

    private JLabel fieldLabel(String text) {
        JLabel l = new JLabel(text);
        l.setForeground(TEXT);
        l.setFont(new Font("Segoe UI", Font.BOLD, 13));
        l.setBorder(new EmptyBorder(0, 0, 6, 0));
        return l;
    }

    private void setupMsgLabel(JLabel l) {
        l.setForeground(DANGER);
        l.setFont(new Font("Segoe UI", Font.PLAIN, 12));
    }

    private void setupTextField(JTextField tf, String placeholder) {
        tf.setBorder(normalBorder);
        tf.setForeground(TEXT);
        tf.setFont(new Font("Segoe UI", Font.PLAIN, 13));
        tf.setBackground(Color.WHITE);
        tf.setToolTipText(placeholder);
    }

    private JButton makeButton(String text, boolean primary) {
        JButton b = new JButton(text);
        b.setFocusPainted(false);
        b.setCursor(Cursor.getPredefinedCursor(Cursor.HAND_CURSOR));
        b.setFont(new Font("Segoe UI", Font.BOLD, 13));
        b.setBorder(new CompoundBorder(
                new LineBorder(primary ? PRIMARY : BORDER, 1, true),
                new EmptyBorder(10, 14, 10, 14)
        ));
        b.setOpaque(true);
        b.setBackground(primary ? PRIMARY : new Color(0xEEF3FD));
        b.setForeground(primary ? Color.WHITE : PRIMARY);

        b.addMouseListener(new MouseAdapter() {
            @Override public void mouseEntered(MouseEvent e) {
                if (primary) b.setBackground(new Color(0x1666D3));
                else b.setBackground(new Color(0xE3ECFF));
            }
            @Override public void mouseExited(MouseEvent e) {
                b.setBackground(primary ? PRIMARY : new Color(0xEEF3FD));
            }
        });
        return b;
    }

    private JPanel resultRow(String label, JLabel value) {
        JPanel row = new JPanel(new BorderLayout(10, 0));
        row.setBackground(CARD_BG);

        JLabel l = new JLabel(label);
        l.setForeground(MUTED);
        l.setFont(new Font("Segoe UI", Font.PLAIN, 13));

        value.setForeground(TEXT);
        value.setFont(new Font("Segoe UI", Font.BOLD, 13));
        value.setHorizontalAlignment(SwingConstants.RIGHT);

        row.add(l, BorderLayout.WEST);
        row.add(value, BorderLayout.EAST);
        return row;
    }

    // ---------------- Validation + Formatting ----------------

    private static class ValidationResult {
        boolean ok;
        String reason;
        ValidationResult(boolean ok, String reason) { this.ok = ok; this.reason = reason; }
    }

    private ValidationResult validateLine(String text) {
        if (text == null) return new ValidationResult(false, "Enter two integers separated by a space.");
        String t = text.trim();
        if (t.isEmpty()) return new ValidationResult(false, "Enter two integers separated by a space.");

        String[] parts = t.split("\\s+");
        if (parts.length < 2) return new ValidationResult(false, "Need 2 numbers (example: 12 6).");
        if (parts.length > 2) return new ValidationResult(false, "Only 2 numbers are allowed on this line.");

        String a = parts[0].replace(",", "");
        String b = parts[1].replace(",", "");

        if (!isIntToken(a) || !isIntToken(b)) return new ValidationResult(false, "Integers only (no letters/decimals).");
        return new ValidationResult(true, "");
    }

    private static boolean isIntToken(String s) {
        if (s == null || s.isEmpty()) return false;
        if (s.startsWith("-")) return s.length() > 1 && s.substring(1).matches("\\d+");
        return s.matches("\\d+");
    }

    private static long parseLongToken(String tok) {
        String raw = tok.replace(",", "");
        return Long.parseLong(raw);
    }

    private static String formatWithCommas(String tok) {
        String raw = tok.replace(",", "");
        if (!isIntToken(raw)) return tok;
        long n = Long.parseLong(raw);
        return NumberFormat.getInstance(Locale.US).format(n);
    }

    private void wireLiveValidation(JTextField field, JLabel reasonLabel) {
        Document doc = field.getDocument();

        doc.addDocumentListener(new DocumentListener() {
            @Override public void insertUpdate(DocumentEvent e) { onChange(); }
            @Override public void removeUpdate(DocumentEvent e) { onChange(); }
            @Override public void changedUpdate(DocumentEvent e) { onChange(); }

            private void onChange() {
                if (formattingInProgress) return;

                String text = field.getText();
                if (text == null) text = "";

                // Auto-format first two tokens with commas.
                // Keep it simple and stable: cursor goes to end after format.
                String formatted = autoFormatLine(text);
                if (!formatted.equals(text)) {
                    formattingInProgress = true;
                    field.setText(formatted);
                    field.setCaretPosition(field.getText().length());
                    formattingInProgress = false;
                }

                // Validate
                String t = field.getText().trim();
                if (t.isEmpty()) {
                    field.setBorder(normalBorder);
                    reasonLabel.setText(" ");
                    return;
                }

                ValidationResult vr = validateLine(field.getText());
                if (vr.ok) {
                    field.setBorder(normalBorder);
                    reasonLabel.setText(" ");
                } else {
                    field.setBorder(errorBorder);
                    reasonLabel.setText(vr.reason);
                }
            }
        });

        field.addFocusListener(new FocusAdapter() {
            @Override public void focusLost(FocusEvent e) {
                // On blur, re-run validation and formatting once.
                if (formattingInProgress) return;
                String text = field.getText();
                String formatted = autoFormatLine(text);
                if (!formatted.equals(text)) {
                    formattingInProgress = true;
                    field.setText(formatted);
                    formattingInProgress = false;
                }
                String t = field.getText().trim();
                if (t.isEmpty()) {
                    field.setBorder(normalBorder);
                    reasonLabel.setText(" ");
                    return;
                }
                ValidationResult vr = validateLine(field.getText());
                field.setBorder(vr.ok ? normalBorder : errorBorder);
                reasonLabel.setText(vr.ok ? " " : vr.reason);
            }
        });
    }

    private String autoFormatLine(String text) {
        String t = text.trim();
        if (t.isEmpty()) return text;

        String[] parts = t.split("\\s+");
        // Format first two tokens; keep extras as-is (validation will catch extras)
        if (parts.length >= 1) parts[0] = formatWithCommas(parts[0]);
        if (parts.length >= 2) parts[1] = formatWithCommas(parts[1]);

        String joined = String.join(" ", parts);
        return joined;
    }

    // ---------------- Actions ----------------

    private void compute() {
        status.setForeground(MUTED);
        status.setText(" ");

        String s1 = in1.getText().trim();
        String s2 = in2.getText().trim();

        ValidationResult v1 = validateLine(s1);
        ValidationResult v2 = validateLine(s2);

        if (s1.isEmpty()) v1 = new ValidationResult(false, "First input is required.");
        if (s2.isEmpty()) v2 = new ValidationResult(false, "Second input is required.");

        msg1.setText(v1.ok ? " " : v1.reason);
        msg2.setText(v2.ok ? " " : v2.reason);
        in1.setBorder(v1.ok ? normalBorder : errorBorder);
        in2.setBorder(v2.ok ? normalBorder : errorBorder);

        if (!v1.ok || !v2.ok) {
            JOptionPane.showMessageDialog(this, (!v1.ok ? v1.reason : v2.reason),
                    "Invalid Input", JOptionPane.ERROR_MESSAGE);
            status.setForeground(DANGER);
            status.setText("Fix the highlighted input(s).");
            return;
        }

        try {
            String[] p1 = s1.split("\\s+");
            String[] p2 = s2.split("\\s+");

            long a = parseLongToken(p1[0]);
            long b = parseLongToken(p1[1]);
            long c = parseLongToken(p2[0]);
            long d = parseLongToken(p2[1]);

            long prod1 = a * b;
            long prod2 = c * d;

            double avg = (a + b + c + d) / 4.0;

            out1.setForeground(OK);
            out2.setForeground(OK);
            out3.setForeground(OK);

            out1.setText(NumberFormat.getInstance(Locale.US).format(prod1));
            out2.setText(NumberFormat.getInstance(Locale.US).format(prod2));

            // Average: if whole number, show as integer with commas; else show decimal.
            if (Math.abs(avg - Math.rint(avg)) < 1e-12) {
                long avgInt = (long) Math.rint(avg);
                out3.setText(NumberFormat.getInstance(Locale.US).format(avgInt));
            } else {
                out3.setText(String.valueOf(avg));
            }

            status.setForeground(OK);
            status.setText("Computed successfully ✔");

        } catch (NumberFormatException ex) {
            JOptionPane.showMessageDialog(this, "Numbers are too large or invalid.",
                    "Invalid Input", JOptionPane.ERROR_MESSAGE);
            status.setForeground(DANGER);
            status.setText("Invalid number format.");
        }
    }

    private void clear() {
        in1.setText("");
        in2.setText("");

        in1.setBorder(normalBorder);
        in2.setBorder(normalBorder);

        msg1.setText(" ");
        msg2.setText(" ");

        out1.setForeground(TEXT);
        out2.setForeground(TEXT);
        out3.setForeground(TEXT);

        out1.setText("—");
        out2.setText("—");
        out3.setText("—");

        status.setForeground(MUTED);
        status.setText(" ");

        in1.requestFocusInWindow();
    }

    // ---------------- Utility ----------------

    private static void setUIFont(Font f) {
        java.util.Enumeration<Object> keys = UIManager.getDefaults().keys();
        while (keys.hasMoreElements()) {
            Object key = keys.nextElement();
            Object value = UIManager.get(key);
            if (value instanceof FontUIResource) {
                UIManager.put(key, new FontUIResource(f));
            }
        }
    }

    public static void main(String[] args) {
        SwingUtilities.invokeLater(() -> {
            // Native feel
            try { UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName()); }
            catch (Exception ignored) {}
            new lab1_1().setVisible(true);
        });
    }
}